### PR TITLE
MVVM Sturcture Implementation

### PR DESCRIPTION
**Pull request by AJ-iOSDev**
CollectionView Diffable datasource with MVVM Sturcture - Inprogress

### CODE REVIEW

File: AppOne/AppOne/Model/Brands.swift [hunk 1]
```diff
@@ -0,0 +1,16 @@
...
+//
+
+struct Brands : Hashable {
+    let name: String
+    let headQuarters: String
+    let ceo: String
+
+    func hash(into hasher: inout Hasher) {
+        hasher.combine(name)
+    }
```
Comment (JoeHoHH) [inline AppOne/AppOne/Model/Brands.swift:15]: can be removed, compiler synthesizes it, also the hash only over the name is not very reliable

File: AppOne/AppOne/Model/SectionTitles.swift [hunk 2]
```diff
@@ -0,0 +1,14 @@
...
+//
+//  Created by Asapu, Ajaikumar on 21.11.22.
+//
+
+struct SectionTitles : Hashable {
+    let title: String
+
+    func hash(into hasher: inout Hasher) {
+        hasher.combine(title)
+    }
```
Comment (JoeHoHH) [inline AppOne/AppOne/Model/SectionTitles.swift:13]: can be removed

File: AppOne/AppOne/View/HomeVC.swift [hunk 3]
```diff
@@ -0,0 +1,39 @@
+//
+//  HomeVC.swift
+//  AppOne
+//
+//  Created by Asapu, Ajaikumar on 19.11.22.
+//
+
+import UIKit
+
+class HomeVC: UIViewController {
```
Comment (JoeHoHH) [inline AppOne/AppOne/View/HomeVC.swift:10]: final

File: AppOne/AppOne/View/HomeVC.swift [hunk 4]
```diff
@@ -0,0 +1,39 @@
...
+    private var brandsCollectionView: UICollectionView!
+    private var dataSource: UICollectionViewDiffableDataSource<SectionTitles, Brands>!
```
Comment (JoeHoHH) [inline AppOne/AppOne/View/HomeVC.swift:12]: dont use force unwrap

File: AppOne/AppOne/Model/SectionTitles.swift [hunk 5]
```diff
@@ -0,0 +1,14 @@
+//
+//  SectionTitles.swift
+//  AppOne
+//
+//  Created by Asapu, Ajaikumar on 21.11.22.
+//
+
+struct SectionTitles : Hashable {
```
Comment (JoeHoHH) [inline AppOne/AppOne/Model/SectionTitles.swift:8]: singular `SectionTitle`
```suggestion
struct SectionTitle : Hashable {
```

File: AppOne/AppOne/View/HomeVC.swift [hunk 6]
```diff
@@ -0,0 +1,39 @@
...
+    
+    override func viewDidLoad() {
+        super.viewDidLoad()
+        configureCollectionView()
+        configureDataSource()
+    }
+
+    func configureCollectionView() {
+        brandsCollectionView = UICollectionView(frame: view.bounds, collectionViewLayout: configureLayout())
```
Comment (JoeHoHH) [inline AppOne/AppOne/View/HomeVC.swift:21]: this should get it"s size over auto layout

File: AppOne/AppOne/View/HomeVC.swift [hunk 7]
```diff
@@ -0,0 +1,39 @@
...
+        brandsCollectionView.register(UICollectionViewCell.self, forCellWithReuseIdentifier: "Cell")
```
Comment (JoeHoHH) [inline AppOne/AppOne/View/HomeVC.swift:22]: this is not needed there is a `UICollectionView.CellRegistration`. This is the recommended way to go.

File: AppOne/AppOne/View/HomeVC.swift [hunk 8]
```diff
@@ -0,0 +1,39 @@
...
+        view.addSubview(brandsCollectionView)
+    }
+
+    func configureLayout() -> UICollectionViewLayout {
+        UICollectionViewCompositionalLayout.list(using: UICollectionLayoutListConfiguration(appearance: .grouped))
+    }
+
+    func configureDataSource() {
+        dataSource = UICollectionViewDiffableDataSource(collectionView: brandsCollectionView, cellProvider: { collectionView, indexPath, itemIdentifier in
+            let cell = collectionView.dequeueReusableCell(withReuseIdentifier: "Cell", for: indexPath)
```
Comment (JoeHoHH) [inline AppOne/AppOne/View/HomeVC.swift:32]: `collectionView.dequeueConfiguredReusableCell(using: cellReg, for: indexPath, item: itemIdentifier)`
and then set the items content in the cellreg

**Comment by JoeHoHH** [comment]
You completly missed to fill the datasource. so you need to have a snapshot and fill it with the section and the data. Have a look at `NSDiffableDataSnapshot`

File: AppOne/AppOne/View/HomeVC.swift (hunk 7 above)
Comment (AJ-iOSDev) [reply AppOne/AppOne/View/HomeVC.swift:22]: Ok

File: AppOne/AppOne/Model/Brands.swift (hunk 1 above)
Comment (JoeHoHH) [inline AppOne/AppOne/Model/Brands.swift:15]: not sure why you deleted this 😆

File: AppOne/AppOne/HomeVC.swift [hunk 9]
```diff
@@ -1,15 +0,0 @@
...
-//
-
-import UIKit
-
-class HomeVC: UIViewController {
-    override func viewDidLoad() {
-        super.viewDidLoad()
-        print("HomeVC")
-    }
-}
```
Comment (JoeHoHH) [inline AppOne/AppOne/HomeVC.swift:15]: Why is this deleted? somehow it is not the version from main.....

**Review by JoeHoHH — DISMISSED** [review]

File: AppOne/AppOne/Model/Brands.swift [hunk 10]
```diff
@@ -0,0 +1,16 @@
+//
+//  Brands.swift
+//  AppOne
+//
+//  Created by Asapu, Ajaikumar on 20.11.22.
+//
+
+struct Brands : Hashable {
```
Comment (JoeHoHH) [inline AppOne/AppOne/Model/Brands.swift:8]:
```suggestion
struct Brand : Hashable {
```

File: AppOne/AppOne/View/HomeVC.swift [hunk 11]
```diff
@@ -0,0 +1,52 @@
...
+    }
+
+//    func configureLayout() -> UICollectionViewLayout {
+//        UICollectionViewCompositionalLayout.list(using: UICollectionLayoutListConfiguration(appearance: .plain))
+//    }
+
+    func configureDataSource() {
+        let cellRegistration = UICollectionView.CellRegistration<SectionTitles, Brands> { cell, indexPath, item in
+            var contentConfiguration = cell.defaultContentConfiguration()
+            contentConfiguration.text = "\(item)"
```
Comment (JoeHoHH) [inline AppOne/AppOne/View/HomeVC.swift:34]: here you should use `item.name`

File: AppOne/AppOne/View/HomeVC.swift [hunk 12]
```diff
@@ -0,0 +1,52 @@
...
+            (collectionView: UICollectionView, indexPath: IndexPath, itemIdentifier: Int) -> UICollectionViewCell? in
+
+            return collectionView.dequeueConfiguredReusableCell(using: cellRegistration,
+                                                                for: indexPath,
+                                                                item: itemIdentifier)
+        }
+    }
+
+    func applySnapshot() {
+        var snapshot = NSDiffableDataSourceSnapshot<SectionTitles, Brands>()
```
Comment (JoeHoHH) [inline AppOne/AppOne/View/HomeVC.swift:49]: you still dont fill the snapshot

Comment (JoeHoHH) [reply AppOne/AppOne/View/HomeVC.swift:49]: go with  ```
snapshot.appendSections([some Sectiontitle])
        snapshot.appendItems([Brand])
```

File: AppOne/AppOne/View/HomeVC.swift [hunk 13]
```diff
@@ -0,0 +1,52 @@
...
+    
+    override func viewDidLoad() {
+        super.viewDidLoad()
+      //  configureCollectionView()
+        configureDataSource()
+        applySnapshot()
+    }
+
+    func configureCollectionView() {
+//        brandsCollectionView = UICollectionView(frame: view.bounds, collectionViewLayout: configureLayout())
```
Comment (JoeHoHH) [inline AppOne/AppOne/View/HomeVC.swift:22]: we somehow have to let the collectionView know it's size

File: AppOne/AppOne/SceneDelegate.swift [hunk 14]
```diff
@@ -16,8 +16,8 @@ class SceneDelegate: UIResponder, UIWindowSceneDelegate {
         // Use this method to optionally configure and attach the UIWindow `window` to the provided UIWindowScene `scene`.
         // If using a storyboard, the `window` property will automatically be initialized and attached to the scene.
         // This delegate does not imply the connecting scene or session are new (see `application:configurationForConnectingSceneSession` instead).
-        guard let _ = (scene as? UIWindowScene) else { return }
-        window = UIWindow()
+        guard let windowscene = (scene as? UIWindowScene) else { return }
```
Comment (JoeHoHH) [inline AppOne/AppOne/SceneDelegate.swift:19]:
```suggestion
        guard let windowscene = scene as? UIWindowScene else { return }
```

Comment (JoeHoHH) [inline AppOne/AppOne/SceneDelegate.swift:19]:
```suggestion
        guard let windowScene = (scene as? UIWindowScene) else { return }
```

File: AppOne/AppOne/View/HomeVC.swift [hunk 15]
```diff
@@ -0,0 +1,53 @@
...
+//
+
+import UIKit
+
+final class HomeVC: UIViewController {
+    private lazy var brandsCollection: UICollectionView = {
+        let collectionV = UICollectionView(
+            frame: view.bounds,
+            collectionViewLayout: UICollectionViewCompositionalLayout.list(using: UICollectionLayoutListConfiguration(appearance: .plain)))
+        return collectionV
```
Comment (JoeHoHH) [inline AppOne/AppOne/View/HomeVC.swift:15]: you miss the `translatesAutoresizingMaskIntoConstraints`

File: AppOne/AppOne/View/HomeVC.swift [hunk 16]
```diff
@@ -0,0 +1,53 @@
...
+//
+//  Created by Asapu, Ajaikumar on 19.11.22.
+//
+
+import UIKit
+
+final class HomeVC: UIViewController {
+    private lazy var brandsCollection: UICollectionView = {
+        let collectionV = UICollectionView(
+            frame: view.bounds,
```
Comment (JoeHoHH) [inline AppOne/AppOne/View/HomeVC.swift:13]: should be set to .zero to let Auto Layout do the sizing

File: AppOne/AppOne/View/HomeVC.swift [hunk 17]
```diff
@@ -0,0 +1,53 @@
...
+            collectionViewLayout: UICollectionViewCompositionalLayout.list(using: UICollectionLayoutListConfiguration(appearance: .plain)))
+        return collectionV
+    }()
+
+    private lazy var dataSource: UICollectionViewDiffableDataSource<SectionTitles, Brands> = {
+        let cellRegistration = UICollectionView.CellRegistration<UICollectionViewListCell, Brands> { cell, index, res  in
+            var contentConfig = cell.defaultContentConfiguration()
+            contentConfig.text = res.name
```
Comment (JoeHoHH) [inline AppOne/AppOne/View/HomeVC.swift:21]: here you could also set image and /or secondary text if needed

File: AppOne/AppOne/View/HomeVC.swift [hunk 18]
```diff
@@ -0,0 +1,53 @@
...
+
+        return UICollectionViewDiffableDataSource(collectionView: brandsCollection) { collectionView, indexPath, itemIdentifier in
+            collectionView.dequeueConfiguredReusableCell(using: cellRegistration, for: indexPath, item: itemIdentifier)
+        }
+    }()
+    
+    override func viewDidLoad() {
+        super.viewDidLoad()
+        view.addSubview(brandsCollection)
+        brandsCollection.translatesAutoresizingMaskIntoConstraints = false
```
Comment (JoeHoHH) [inline AppOne/AppOne/View/HomeVC.swift:33]: should move up no need to split. whole config code for ui component would be in one modular place then

File: AppOne/AppOne/View/HomeVC.swift [hunk 19]
```diff
@@ -0,0 +1,53 @@
...
+            brandsCollection.leadingAnchor.constraint(equalTo: view.leadingAnchor),
+            brandsCollection.trailingAnchor.constraint(equalTo: view.trailingAnchor),
+            brandsCollection.topAnchor.constraint(equalTo: view.topAnchor),
+            brandsCollection.bottomAnchor.constraint(equalTo: view.bottomAnchor)
+        ])
+    }
+
+    private func applySnapshot(sec: SectionTitles, bran: Brands) {
+        var snapshot = NSDiffableDataSourceSnapshot<SectionTitles, Brands>()
+        snapshot.appendSections([sec])
```
Comment (JoeHoHH) [inline AppOne/AppOne/View/HomeVC.swift:49]: it is correct here with the current code, but we could make sections an enum and let it conform to `CaseIterable`, but we can discuss this later.

**Review by JoeHoHH — CHANGES_REQUESTED** [review]

**Comment by JoeHoHH** [comment]
Good job, just some minor things. Almost there. 😄

File: AppOne/AppOne/Model/SectionTitles.swift (hunk 2 above)
Comment (AJ-iOSDev) [reply AppOne/AppOne/Model/SectionTitles.swift:13]: don't we need hasher ever here?

File: AppOne/AppOne/View/HomeVC.swift (hunk 15 above)
Comment (AJ-iOSDev) [reply AppOne/AppOne/View/HomeVC.swift:15]: Ok moved it to here

File: AppOne/AppOne/View/HomeVC.swift (hunk 16 above)
Comment (AJ-iOSDev) [reply AppOne/AppOne/View/HomeVC.swift:13]: Ok

File: AppOne/AppOne/View/HomeVC.swift (hunk 19 above)
Comment (AJ-iOSDev) [reply AppOne/AppOne/View/HomeVC.swift:49]: Ok

**Comment by AJ-iOSDev** [comment]
tried playing with custom image and more rows in collectionview. next steps would be that and proper data handling by removing static data in viewdidload

**Comment by AJ-iOSDev** [comment]
and also having doubt why would we remove hasher

**Comment by AJ-iOSDev** [comment]
Doubts: Don't understand how append sections work here.
/// Metatypes such as `Int.Type` implicitly conform to the `Sendable` protocol.
public protocol Sendable {
}

If datasource gets only item data then shouldn't we send data in sections? only int have to send (kind of no.of sections)?

File: AppOne/AppOne/Model/SectionTitles.swift (hunk 2 above)
Comment (JoeHoHH) [reply AppOne/AppOne/Model/SectionTitles.swift]: If you conform to Hashable and all properties conform to Hashabale, then you get the default implementation over all properties for free from the compiler

**Comment by JoeHoHH** [comment]
I dont understand your question for Sendable? Why is this connected to you?

File: AppOne/AppOne/View/HomeVC.swift [hunk 20]
```diff
@@ -0,0 +1,57 @@
...
+//
+//  Created by Asapu, Ajaikumar on 19.11.22.
+//
+
+import UIKit
+
+final class HomeVC: UIViewController {
+    private lazy var brandsCollection: UICollectionView = {
+        let collectionV = UICollectionView(
+            frame: view.bounds,
```
Comment (JoeHoHH) [inline AppOne/AppOne/View/HomeVC.swift:13]: still we dont use `.zero` here

File: AppOne/AppOne/View/HomeVC.swift [hunk 21]
```diff
@@ -0,0 +1,57 @@
...
+    }()
+
+    private lazy var homeVC = HomeViewModel()
+
+    private lazy var dataSource: UICollectionViewDiffableDataSource<SectionTitle, Brand> = {
+        let cellRegistration = UICollectionView.CellRegistration<UICollectionViewListCell, Brand> { cell, index, res  in
+            var contentConfig = cell.defaultContentConfiguration()
+            contentConfig.text = res.name
+            contentConfig.secondaryText = res.headQuarters
+//            contentConfig.image = UIImage(named: "Flower.png")
```
Comment (JoeHoHH) [inline AppOne/AppOne/View/HomeVC.swift:26]: why did you comment it here? should work. right?

Comment (AJ-iOSDev) [reply AppOne/AppOne/View/HomeVC.swift:26]: no not working, only system image working but custom image shows too big

File: AppOne/AppOne/View/HomeVC.swift (hunk 19 above)
Comment (JoeHoHH) [reply AppOne/AppOne/View/HomeVC.swift]: You can try an enum and conform to `CaseIterable`

**Comment by JoeHoHH** [comment]
> Doubts: Don't understand how append sections work here. /// Metatypes such as `Int.Type` implicitly conform to the `Sendable` protocol. public protocol Sendable { }
> 
> If datasource gets only item data then shouldn't we send data in sections? only int have to send (kind of no.of sections)?

What is your doubt? Normally it is straightforward, a Collectionview has Sections + items in that section. Also it is possible to have 1...n sections. Then you just give the datasource the sections and then you apply the data per section. Not sure where your doubt/misunderstanding is?

**Review by JoeHoHH — DISMISSED** [review]

File: AppOne/AppOne/View/HomeVC.swift (hunk 21 above)
Comment (JoeHoHH) [reply AppOne/AppOne/View/HomeVC.swift:26]: that can be changed over imageprovider

**Review by JoeHoHH — APPROVED** [review]
